docs/conf.py

```diff
@@ -273,8 +273,10 @@ def generate_r_docs(app: Sphinx) -> None:
             r-base=4.1.0=hb67fd72_2 \
             r-data.table=1.14.0=r41hcfec24a_0 \
             r-jsonlite=1.7.2=r41hcfec24a_0 \
+            r-knitr=1.35=r41hc72bb7e_0 \
             r-matrix=1.3_4=r41he454529_0 \
             r-pkgdown=1.6.1=r41hc72bb7e_0 \
+            r-rmarkdown=2.11=r41hc72bb7e_0 \
```

Comment:
We can't check these changes at RTD because this PR is made from a fork. I suggest re-target this PR's branch to new let's say `dev` branch and merge it instantly. Then I'll be able to activate RTD test builds for that branch. After all checks passed we merge `dev` into `master` via new PR. WDYT about this plan?

Also, I guess we need some changes in [our `pkgdown` config file](https://github.com/microsoft/LightGBM/blob/master/R-package/pkgdown/_pkgdown.yml) to make vignettes publicly available and rendered at our docs.

Reply:
>  suggest re-target this PR's branch to new let's say dev branch and merge it instantly

Yep I agree! Good idea. I'll make that `dev` branch right now and re-target this.

> I guess we need some changes in our pkgdown config file

Ah you're right! According to https://pkgdown.r-lib.org/articles/pkgdown.html#articles, `{pkgdown}` always automatically builds the HTML for vignettes, but I think since `{lightgbm}`'s config explicitly sets the navigation, we need to explicitly add a navbar item for it.

Added in https://github.com/microsoft/LightGBM/pull/3946/commits/ddd9a62b1fda06c12f89c87b4d5a493f59909a29.

Tested locally with

```shell
sh build-cran-package.sh
R CMD INSTALL --with-keep.source lightgbm_3.3.1.99.tar.gz
cd R-package
Rscript -e "pkgdown::build_site( \
            lazy = FALSE \
            , install = FALSE \
            , devel = FALSE \
            , examples = TRUE \
            , run_dont_run = TRUE \
            , seed = 42L \
            , preview = FALSE \
            , new_process = TRUE \
        )
        "
open docs/index.html
```

Vignettes end up on an "Articles" page like this:

![image](https://user-images.githubusercontent.com/7608904/140613984-a1a0094f-be73-43ba-a90e-e8f76c8949fd.png)

And then clicking through the link there, I see

![image](https://user-images.githubusercontent.com/7608904/140614070-bb0cca6a-77db-4bc7-97b0-38161216cb29.png)

Reply:
I just pointed this at a new LightGBM branch, `dev-vignettes`. Decided not to literally use "`dev`", just because that has a [special meaning in some git workflows](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow), and I didn't want anyone coming to this repo to misinterpret and start building from that branch.

Reply:
moved this over to https://github.com/microsoft/LightGBM/pull/4775